### PR TITLE
Store promises with more metadata

### PR DESCRIPTION
**Pull request by Waldz**
Extended promises from accountant with more metadata:
- identity, hermesID, channelID - describes which identity and with who has with opened payment channel
- it allows to list&filter all active payment channels without knowing hermes IDs

### CODE REVIEW

File: session/pingpong/hermes_promise_settler.go [hunk 1]
```diff
@@ -185,6 +185,15 @@ func (aps *hermesPromiseSettler) resyncState(id identity.Identity, hermesID comm
 	return nil
 }
 
+func (aps *hermesPromiseSettler) getLastPromise(id identity.Identity, hermesID common.Address) (HermesPromise, error) {
+	channelID, err := crypto.GenerateProviderChannelID(id.Address, hermesID.Hex())
+	if err != nil {
+		return HermesPromise{}, errors.Wrap(err, "could not generate provider channel address")
```
Comment (soffokl) [inline session/pingpong/hermes_promise_settler.go:191]:
```suggestion
		return HermesPromise{}, fmt.Errorf("could not generate provider channel address: %w", err)
```

**Review by soffokl — APPROVED** [review]

File: session/pingpong/hermes_promise_storage.go [hunk 2]
```diff
@@ -71,22 +77,12 @@ func (aps *HermesPromiseStorage) Store(id identity.Identity, hermesID common.Add
 		return ErrAttemptToOverwrite
 	}
 
-	channel, err := crypto.GenerateProviderChannelID(id.Address, hermesID.Hex())
-	if err != nil {
-		return errors.Wrap(err, "could not generate provider channel address")
-	}
-
-	return errors.Wrap(aps.bolt.SetValue(hermesPromiseBucketName, channel, promise), "could not store hermes promise")
+	return errors.Wrap(aps.bolt.SetValue(hermesPromiseBucketName, promise.ChannelID, promise), "could not store hermes promise")
```
Comment (zolia) [inline session/pingpong/hermes_promise_storage.go:80]: we can use fmt.Errorf with %w here.

File: session/pingpong/hermes_promise_storage.go [hunk 3]
```diff
@@ -97,9 +93,55 @@ func (aps *HermesPromiseStorage) get(id identity.Identity, hermesID common.Addre
...
+// List fetches the promise for the given hermes.
+func (aps *HermesPromiseStorage) List(filter HermesPromiseFilter) ([]HermesPromise, error) {
+	aps.lock.Lock()
+	defer aps.lock.Unlock()
+
+	result := make([]HermesPromise, 0)
+	err := aps.bolt.DB().Bolt.View(func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket([]byte(hermesPromiseBucketName))
+		if bucket == nil {
+			return nil
```
Comment (zolia) [inline session/pingpong/hermes_promise_storage.go:118]: wrap an error

File: session/pingpong/hermes_promise_storage.go [hunk 4]
```diff
@@ -21,10 +21,13 @@ import (
 	"math/big"
 	"sync"
 
+	"github.com/asdine/storm/v3/codec/json"
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/mysteriumnetwork/node/core/storage/boltdb"
 	"github.com/mysteriumnetwork/node/identity"
 	"github.com/mysteriumnetwork/payments/crypto"
 	"github.com/pkg/errors"
+	"go.etcd.io/bbolt"
```
Comment (zolia) [inline session/pingpong/hermes_promise_storage.go:30]: don't use storm?

File: session/pingpong/hermes_promise_storage.go [hunk 5]
```diff
@@ -97,9 +93,55 @@ func (aps *HermesPromiseStorage) get(id identity.Identity, hermesID common.Addre
...
+		}
+
+		return bucket.ForEach(func(k, v []byte) error {
+			if string(k) == "__storm_metadata" {
+				return nil
+			}
+
+			var entry HermesPromise
+			if err := json.Codec.Unmarshal(v, &entry); err != nil {
+				return err
```
Comment (zolia) [inline session/pingpong/hermes_promise_storage.go:128]: wrap an error

**Review by zolia — APPROVED** [review]

File: session/pingpong/hermes_promise_storage.go (hunk 4 above)
Comment (Waldz) [reply session/pingpong/hermes_promise_storage.go:30]: Why not, don't see a point to abstract DB lib as we don't use 2 abstractions. In tests we don't mock DB lib, but rather mock storm, by creating fixtures in it.

File: session/pingpong/hermes_promise_handler.go [hunk 6]
```diff
@@ -178,6 +177,13 @@ func (aph *HermesPromiseHandler) handleNodeStopEvents(e event.Payload) {
 func (aph *HermesPromiseHandler) requestPromise(er enqueuedRequest) {
 	defer func() { close(er.errChan) }()
 
+	providerID := er.providerID
+	hermesID := common.HexToAddress(er.em.HermesID)
+	channelID, err := crypto.GenerateProviderChannelID(providerID.Address, hermesID.Hex())
+	if err != nil {
+		er.errChan <- fmt.Errorf("could not generate provider channel address: %w", err)
+	}
```
Comment (vkuznecovas) [inline session/pingpong/hermes_promise_handler.go:185]: you should probably return here, no point in continuing if you don't have a channel id

**Review by vkuznecovas — CHANGES_REQUESTED** [review]

File: session/pingpong/hermes_promise_handler.go (hunk 6 above)
Comment (Waldz) [reply session/pingpong/hermes_promise_handler.go:186]: nice catch, fixed

**Review by vkuznecovas — APPROVED** [review]

**Pull request merged**